File: app/internal/service/stats_service.go
package service

// 业务编排：统计聚合服务。依赖 ProjectRepository/UnitRepository 接口，可注入 mock。

import (
	"sort"

	"heritage-mgmt/internal/domain"
)

// StatsService 聚合工程统计。
type StatsService struct {
	projects ProjectRepository
	units    UnitRepository
}

// Aggregate 按单位/类型/年份/状态聚合资金与支付，返回与原约定一致的响应体。
func (svc *StatsService) Aggregate() map[string]interface{} {
	projects, _ := svc.projects.ListProjects(0, "", "")
	uname := map[int64]string{}
	units, _ := svc.units.ListUnits()
	for _, u := range units {
		uname[u.ID] = u.Name
	}
	fv := func(v *float64) float64 {
		if v != nil {
			return *v
		}
		return 0
	}
	add := func(m map[string]*domain.StatGroup, key string, p domain.Project) {
		g := m[key]
		if g == nil {
			g = &domain.StatGroup{K: key}
			m[key] = g
		}
		g.N++
		g.Funding += fv(p.CentralFunding)
		g.Paid += fv(p.TotalPaid)
		g.EngC += fv(p.EngContract)
		g.EngP += fv(p.EngPaid)
		g.SupC += fv(p.SupContract)
		g.SupP += fv(p.SupPaid)
		g.DesC += fv(p.DesContract)
		g.DesP += fv(p.DesPaid)
	}
	fin := func(g *domain.StatGroup) { g.Pending = g.Funding - g.Paid }
	toslice := func(m map[string]*domain.StatGroup) []domain.StatGroup {
		out := []domain.StatGroup{}
		for _, g := range m {
			fin(g)
			out = append(out, *g)
		}
		return out
	}
	byUnit := map[string]*domain.StatGroup{}
	byType := map[string]*domain.StatGroup{}
	byYear := map[string]*domain.StatGroup{}
	byStatus := map[string]*domain.StatGroup{}
	var tot domain.StatGroup
	for _, p := range projects {
		add(byUnit, uname[p.UnitID], p)
		tp := p.Ptype
		if tp == "" {
			tp = "其他"
		}
		add(byType, tp, p)
		yr := "未知"
		if len(p.SignDate) >= 4 {
			yr = p.SignDate[:4]
		}
		add(byYear, yr, p)
		st := p.Status
		if st == "" {
			st = "未定"
		}
		add(byStatus, st, p)
		tot.N++
		tot.Funding += fv(p.CentralFunding)
		tot.Paid += fv(p.TotalPaid)
		tot.EngC += fv(p.EngContract)
		tot.EngP += fv(p.EngPaid)
		tot.SupC += fv(p.SupContract)
		tot.SupP += fv(p.SupPaid)
		tot.DesC += fv(p.DesContract)
		tot.DesP += fv(p.DesPaid)
	}
	tot.Pending = tot.Funding - tot.Paid
	us := toslice(byUnit)
	sort.Slice(us, func(i, j int) bool { return us[i].Funding > us[j].Funding })
	ts := toslice(byType)
	sort.Slice(ts, func(i, j int) bool { return ts[i].Funding > ts[j].Funding })
	ys := toslice(byYear)
	sort.Slice(ys, func(i, j int) bool { return ys[i].K < ys[j].K })
	ss := toslice(byStatus)
	return map[string]interface{}{
		"total":     tot,
		"by_unit":   us,
		"by_type":   ts,
		"by_year":   ys,
		"by_status": ss,
	}
}

File: desktop/src/core/stats/StatsService.cpp
#include "StatsService.h"

#include <QHash>
#include <algorithm>

namespace heritage {

namespace {
// optional<double> 取值，NULL→0（与 Go fv 一致）
double fv(const std::optional<double>& v) { return v.has_value() ? *v : 0.0; }

// 累加到分组
void add(StatGroup& g, const Project& p) {
    g.n++;
    g.funding += fv(p.centralFunding);
    g.paid += fv(p.totalPaid);
    g.engC += fv(p.engContract);
    g.engP += fv(p.engPaid);
    g.supC += fv(p.supContract);
    g.supP += fv(p.supPaid);
    g.desC += fv(p.desContract);
    g.desP += fv(p.desPaid);
}

// 把 hash 转为已 finalize(pending) 的有序 vector
QVector<StatGroup> toSlice(QHash<QString, StatGroup>& m, bool sortByFundingDesc) {
    QVector<StatGroup> out;
    out.reserve(m.size());
    for (auto it = m.begin(); it != m.end(); ++it) {
        it->pending = it->funding - it->paid;
        out.append(it.value());
    }
    if (sortByFundingDesc)
        std::sort(out.begin(), out.end(),
                  [](const StatGroup& a, const StatGroup& b) { return a.funding > b.funding; });
    return out;
}
} // namespace

StatsService::StatsService(ProjectRepo& projects, UnitRepo& units) : projects_(projects), units_(units) {}

StatsResult StatsService::aggregate() {
    const QVector<Project> projects = projects_.list(0, QString(), QString());
    QHash<qint64, QString> uname;
    const QVector<Unit> units = units_.list();
    for (const Unit& u : units)
        uname.insert(u.id, u.name);

    QHash<QString, StatGroup> byUnit, byType, byYear, byStatus;
    StatGroup tot;
    // accum：取/建分组并写入 key（与 Go `g := &StatGroup{K: key}` 一致；空 key 保持空）
    auto accum = [](QHash<QString, StatGroup>& m, const QString& key, const Project& p) {
        StatGroup& g = m[key];
        if (g.k.isEmpty())
            g.k = key;
        add(g, p);
    };
    for (const Project& p : projects) {
        accum(byUnit, uname.value(p.unitId), p); // 单位名作 key，可能为空串
        QString tp = p.ptype.isEmpty() ? QStringLiteral("其他") : p.ptype;
        accum(byType, tp, p);
        QString yr = p.signDate.length() >= 4 ? p.signDate.left(4) : QStringLiteral("未知");
        accum(byYear, yr, p);
        QString st = p.status.isEmpty() ? QStringLiteral("未定") : p.status;
        accum(byStatus, st, p);
        add(tot, p); // total 的 k 保持空
    }
    tot.pending = tot.funding - tot.paid;

    StatsResult r;
    r.total = tot;
    r.byUnit = toSlice(byUnit, true);
    r.byType = toSlice(byType, true);
    r.byYear = toSlice(byYear, false);
    std::sort(r.byYear.begin(), r.byYear.end(),
              [](const StatGroup& a, const StatGroup& b) { return a.k < b.k; });
    r.byStatus = toSlice(byStatus, false); // 保持首次出现顺序（Go 用 map 无序，此处取确定性顺序）
    return r;
}

} // namespace heritage
